jre/java/java/util/stream/DoubleStream.java

```diff
@@ -147,19 +147,35 @@ public boolean tryAdvance(DoubleConsumer action) {
   }
 
   static DoubleStream iterate(double seed, DoubleUnaryOperator f) {
+    return iterate(seed, ignore -> true, f);
+  }
+
+  static DoubleStream iterate(double seed, DoublePredicate hasNext, DoubleUnaryOperator f) {
     Spliterator.OfDouble spliterator =
         new Spliterators.AbstractDoubleSpliterator(
             Long.MAX_VALUE, Spliterator.IMMUTABLE | Spliterator.ORDERED) {
+          private boolean first = true;
           private double next = seed;
+          private boolean terminated = false;
 
           @Override
           public boolean tryAdvance(DoubleConsumer action) {
+            if (terminated) {
```

Comment:
We were not tracking termination before. Was it bug? Can we fix that separately?

Reply:
The old implementation is an infinite stream (the git diff is somewhat confusing here), so it was never possible to terminate. That is, Java 8 shipped with `iterate​(double seed, DoubleUnaryOperator f)`,  and Java 9 added `iterate​(double seed, DoublePredicate hasNext, DoubleUnaryOperator next)`

Technically we could omit the terminated boolean entirely and test both at the beginning of `tryAdvance` and also after incrementing, but I don't think you're suggesting this.

Reply:
Thanks for the clarification, I didn't original API was an infinite stream.